File: tools/sigma-pkg.py
#!/usr/bin/env python3
import os
import sys
import json
import sqlite3

# SigmaOS — sigma-pkg: Sovereign Package Manager
# USP: No external dependencies beyond Python3/SQLite3, hardware-native shard registry.

REGISTRY_PATH = "suites/S36_SovereignPkg/registry.json"
DB_PATH = "suites/S36_SovereignPkg/pkg_graph.db"

def init_db():
    conn = sqlite3.connect(DB_PATH)
    c = conn.cursor()
    c.execute('''CREATE TABLE IF NOT EXISTS packages 
                 (name TEXT PRIMARY KEY, version TEXT, status TEXT)''')
    c.execute('''CREATE TABLE IF NOT EXISTS dependencies 
                 (package TEXT, dependency TEXT, FOREIGN KEY(package) REFERENCES packages(name))''')
    conn.commit()
    return conn

def load_registry():
    if not os.path.exists(REGISTRY_PATH):
        return {"packages": {}}
    with open(REGISTRY_PATH, 'r') as f:
        return json.load(f)

def create_snapshot(reason):
    print(f"[sigma-pkg] Safety: Creating pre-{reason} snapshot...")
    # Call sigma-snap (mocked for now, but linked to the logic)
    snapshot_id = 42 # Mock ID
    print(f"[✓] Snapshot {snapshot_id} created successfully.")
    return snapshot_id

def install(pkg_name):
    create_snapshot("install")
    print(f"[sigma-pkg] Attempting to install shard: {pkg_name}...")
    registry = load_registry()
    if pkg_name not in registry["packages"]:
        print(f"[ERROR] Shard '{pkg_name}' not found in sovereign registry.")
        return False
    
    pkg_data = registry["packages"][pkg_name]
    deps = pkg_data.get("dependencies", [])
    
    conn = init_db()
    c = conn.cursor()
    
    # Resolve dependencies
    for dep in deps:
        print(f"[sigma-pkg] Resolving dependency: {dep}")
        c.execute("SELECT status FROM packages WHERE name=?", (dep,))
        res = c.fetchone()
        if not res or res[0] != "installed":
            if not install(dep):
                print(f"[ERROR] Failed to satisfy dependency '{dep}' for '{pkg_name}'.")
                return False

    # Simulate installation (linking shard)
    print(f"[sigma-pkg] Linking shard {pkg_name} into the lattice...")
    c.execute("INSERT OR REPLACE INTO packages (name, version, status) VALUES (?, ?, ?)", 
              (pkg_name, pkg_data.get("version", "1.0.0"), "installed"))
    for dep in deps:
        c.execute("INSERT OR IGNORE INTO dependencies (package, dependency) VALUES (?, ?)", 
                  (pkg_name, dep))
    conn.commit()
    print(f"[✓] Shard '{pkg_name}' installed successfully.")
    return True

def remove(pkg_name):
    create_snapshot("remove")
    print(f"[sigma-pkg] Removing shard: {pkg_name}...")
    conn = init_db()
    c = conn.cursor()
    c.execute("DELETE FROM packages WHERE name=?", (pkg_name,))
    c.execute("DELETE FROM dependencies WHERE package=?", (pkg_name,))
    conn.commit()
    print(f"[✓] Shard '{pkg_name}' purged from lattice.")

def list_installed():
    conn = init_db()
    c = conn.cursor()
    c.execute("SELECT name, version FROM packages WHERE status='installed'")
    rows = c.fetchall()
    print("\n--- Installed Sovereign Shards ---")
    for row in rows:
        print(f"  {row[0]} (v{row[1]})")
    print("----------------------------------\n")

def main():
    if len(sys.argv) < 2:
        print("Usage: sigma-pkg <install|remove|list> [shard_name]")
        sys.exit(1)
    
    cmd = sys.argv[1]
    if cmd == "install" and len(sys.argv) > 2:
        install(sys.argv[2])
    elif cmd == "remove" and len(sys.argv) > 2:
        remove(sys.argv[2])
    elif cmd == "list":
        list_installed()
    else:
        print(f"[ERROR] Unknown command or missing arguments: {cmd}")

if __name__ == "__main__":
    main()

File: tools/sigma-pkg.js
import fs from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Paths
const REGISTRY_PATH = path.join(__dirname, '..', 'suites', 'S36_SovereignPkg', 'registry.json');
const DB_PATH = path.join(__dirname, '..', 'suites', 'S36_SovereignPkg', 'pkg_graph.json');

// Helper to ensure target directories exist
function ensureDirs() {
    const registryDir = path.dirname(REGISTRY_PATH);
    if (!fs.existsSync(registryDir)) {
        fs.mkdirSync(registryDir, { recursive: true });
    }
}

// Load Registry
function loadRegistry() {
    ensureDirs();
    if (!fs.existsSync(REGISTRY_PATH)) {
        // Seed default registry if not present
        const defaultReg = {
            packages: {
                "sentinel": { version: "1.0.0", dependencies: ["sovereign-net"] },
                "sovereign-net": { version: "1.2.0", dependencies: [] },
                "zenith-ui": { version: "2.1.0", dependencies: ["sentinel"] }
            }
        };
        fs.writeFileSync(REGISTRY_PATH, JSON.stringify(defaultReg, null, 2));
        return defaultReg;
    }
    return JSON.parse(fs.readFileSync(REGISTRY_PATH, 'utf8'));
}

// Load JSON Database
function loadDB() {
    ensureDirs();
    if (!fs.existsSync(DB_PATH)) {
        const defaultDB = { packages: {}, dependencies: {} };
        fs.writeFileSync(DB_PATH, JSON.stringify(defaultDB, null, 2));
        return defaultDB;
    }
    return JSON.parse(fs.readFileSync(DB_PATH, 'utf8'));
}

// Save JSON Database
function saveDB(db) {
    fs.writeFileSync(DB_PATH, JSON.stringify(db, null, 2));
}

// Create snapshot mock
function createSnapshot(reason) {
    console.log(`[sigma-pkg] Safety: Creating pre-${reason} snapshot...`);
    const snapshotId = Math.floor(Math.random() * 1000) + 1;
    console.log(`[✓] Snapshot ${snapshotId} created successfully.`);
    return snapshotId;
}

// Install package
function install(pkgName) {
    createSnapshot("install");
    console.log(`[sigma-pkg] Attempting to install shard: ${pkgName}...`);
    
    const registry = loadRegistry();
    if (!registry.packages[pkgName]) {
        console.error(`[ERROR] Shard '${pkgName}' not found in sovereign registry.`);
        return false;
    }

    const pkgData = registry.packages[pkgName];
    const deps = pkgData.dependencies || [];

    const db = loadDB();

    // Resolve dependencies
    for (const dep of deps) {
        console.log(`[sigma-pkg] Resolving dependency: ${dep}`);
        if (!db.packages[dep] || db.packages[dep].status !== "installed") {
            if (!install(dep)) {
                console.error(`[ERROR] Failed to satisfy dependency '${dep}' for '${pkgName}'.`);
                return false;
            }
        }
    }

    // Update database (reload since recursive calls might have updated it)
    const activeDB = loadDB();
    activeDB.packages[pkgName] = {
        version: pkgData.version || "1.0.0",
        status: "installed"
    };
    activeDB.dependencies[pkgName] = deps;
    saveDB(activeDB);

    console.log(`[sigma-pkg] Linking shard ${pkgName} into the lattice...`);
    console.log(`[✓] Shard '${pkgName}' installed successfully.`);
    return true;
}

// Remove package
function remove(pkgName) {
    createSnapshot("remove");
    console.log(`[sigma-pkg] Removing shard: ${pkgName}...`);
    
    const db = loadDB();
    if (db.packages[pkgName]) {
        delete db.packages[pkgName];
    }
    if (db.dependencies[pkgName]) {
        delete db.dependencies[pkgName];
    }
    saveDB(db);
    
    console.log(`[✓] Shard '${pkgName}' purged from lattice.`);
}

// List installed packages
function listInstalled() {
    const db = loadDB();
    console.log("\n--- Installed Sovereign Shards ---");
    const pkgs = Object.keys(db.packages);
    if (pkgs.length === 0) {
        console.log("  No shards installed.");
    } else {
        for (const pkg of pkgs) {
            if (db.packages[pkg].status === "installed") {
                console.log(`  ${pkg} (v${db.packages[pkg].version})`);
            }
        }
    }
    console.log("----------------------------------\n");
}

function main() {
    const args = process.argv.slice(2);
    if (args.length < 1) {
        console.log("Usage: node tools/sigma-pkg.js <install|remove|list> [shard_name]");
        process.exit(1);
    }

    const cmd = args[0];
    if (cmd === "install" && args.length > 1) {
        install(args[1]);
    } else if (cmd === "remove" && args.length > 1) {
        remove(args[1]);
    } else if (cmd === "list") {
        listInstalled();
    } else {
        console.error(`[ERROR] Unknown command or missing arguments: ${cmd}`);
    }
}

main();
